orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -52,17 +52,34 @@ TrainingRunner::TrainingRunner(Parameters params, const Environment& env, Sessio
       params_(params),
       session_options_(session_options),
       session_(session_options, env),
-      input_allocator_(params.input_allocator ? params.input_allocator : TrainingUtil::GetCpuAllocator()) {
+      input_allocator_(params.input_allocator ? params.input_allocator : TrainingUtil::GetCpuAllocator()),
+      pipeline_schedule_(static_cast<int>(params_.num_pipeline_stages)),
+      pipeline_worker_pool_(params.num_pipeline_stages) {
   ORT_ENFORCE(!params_.model_path.empty());
   if (!params.weights_to_train.empty())
     ORT_ENFORCE(params.weights_not_to_train.empty());
   ORT_ENFORCE(!params_.training_optimizer_name.empty());
   if (params.partition_optimizer)
     ORT_ENFORCE(params.use_nccl, "Optimizer partitioning is only supported with NCCL distributed training.");
+  
+  if (params_.num_pipeline_stages > 1) {
+    pipeline_context_.pipeline_stage_id = params_.mpi_context.world_rank;
+    pipeline_context_.num_pipeline_stages = params_.num_pipeline_stages;
+    pipeline_context_.num_pipeline_batches = params_.gradient_accumulation_steps - 1;
```

Comment:
>params_.gradient_accumulation_steps - 1; [](start = 45, length = 40)

why?

Reply:
Changed to 
```
pipeline_context_.num_pipeline_batches = params_.gradient_accumulation_steps -1
```
I used the last batch to run model update, but that update can be merged into normal pipeline.